contents/blog/python-client/index.mdx

```diff
@@ -0,0 +1,128 @@
+
+### Python Client Data Structures
+
+The core structures of the client organize metadata about the foundational objects of the OpenLineage spec: runs, jobs and datasets.
+
+A `dataset` is a class consisting of a `name`, `namespace` and, optionally, `facets` array:
+
+```
+@attr.s
+class Dataset:
+	namespace: str = attr.ib()
+	name: str - attr.ib()
+	Facets: Dict = attr.ib(factory=dict)
+```
+
+The same goes for a `job`:
+
+```
+@attr.s
+class Job:
+	namespace: str = attr.ib()
+	name: str - attr.ib()
+	Facets: Dict = attr.ib(factory=dict)
+```
+
+A `RunEvent` sends the time, state, job, run, producer, input and output information needed to construct an OpenLineage job run event:
+
+```
+@attr.s 
+class RunEvent:
+	eventType: RunState = attr.ib(validator=attr.validators.in_(RunState))
+	eventTime: str = attr.ib()
+	run: Run = attr.ib()
+	job: Job = attr.ib()
+	producer: str = attr.ib()
+	inputs: Optional[List[Dataset]] = attr.ib(factory=list)
+	outputs: Optional[List[Dataset]] = attr.ib(factory=list)
+```
+### The OpenLineage-dbt Integration
+
+At a high level, the [dbt integration](https://github.com/OpenLineage/OpenLineage/tree/main/integration/dbt) uses the Python client to push metadata to the OpenLineage backend. The metadata it makes available comprises the run lifecycle, including any dataset inputs and outputs accessed during a job run.
+
+In the `dbt-ol` [script](https://github.com/OpenLineage/OpenLineage/blob/main/integration/dbt/scripts/dbt-ol), the integration uses the project’s `ParentRunMetadata` and `DbtArtifactProcessor` classes, both of which can be found in the OpenLineage common integration, to parse metadata from the dbt `manifest` and `run_result` in order to produce OpenLineage events:  
+
+```
+from openlineage.common.provider.dbt import DbtArtifactProcessor, ParentRunMetadata
+
+#…
+
+if parent_id:
+	parent_namespace, parent_job_name, parent_run_id = parent_id.split(‘/’)
+	parent_run_metadata = ParentRunMetadata(
+		run_id=parent_run_id,
+		job_name=parent_job_name,
+		job_namespace=parent_namespace
+	)
+
+processor = DbtArtifactProcessor(
+producer=PRODUCER,
+target=target,
+job_namespace=job_namespace,
+project_dir=project_dir,
+profile_name=profile_name,
+logger=logger
+)
+```
+
+The integration uses a wrapper for dbt runs because start and complete events are not emitted until execution concludes:
+
+```
+dbt_run_event = dbt_run_event_start(
+	job_name=f“dbt-run-{processor.project[‘name’]}”,
+	job_namespace=job_namespace,
+	parent_run_metadata=parent_run_metadata
+)
+
+dbt_run_metadata = ParentRunMetadata(
+	run_id=dbt_run_event.run.runId,
+	job_name=dbt_run_event.job.name,
+	job_namespace=dbt_run_event.job.namespace
+)
+
+processor.dbt_run_metadata = dbt_run_metadata
+```
+
+After executing dbt, the script parses the metadata using the processor and emits a run event:
+
+```
+events = processor.parse().events()
+
+client.emit(dbt_run_event_end(
+run_id=dbt_run_metadata.run_id,
```

Comment:
I think we also need to indent this block for readability.

Reply:
thanks for catching!